controllers/git_test.go

```diff
@@ -0,0 +1,595 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/config"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-logr/logr"
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	v1core "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	api "github.com/hybrid-cloud-patterns/patterns-operator/api/v1alpha1"
+)
+
+const (
+	mainReference          plumbing.ReferenceName = "refs/heads/main"
+	originURL              string                 = "https://origin.url"
+	targetURL              string                 = "https://target.url"
+	hashCommitMainHead     string                 = "667679cce3942d3dec754b29d0f97500bba57978"
+	hashCommitAmendedHead  string                 = "6ffb7b8f89075d66fba48c4d0000f8fb52720cf1"
+	hashCommitTestBranch   string                 = "0e34ab1c94a4b588ddea45087e956b22bddfa8a2"
+	hashCommitBugfixBranch string                 = "597db674d31dee964f464d84ee0b4f3797bb06dd"
```

Comment:
Can we do some tests with branches and tags please?
I seem to recall one of the popular git modules was rather stupid in this respect.

Reply:
Sure thing.